firedrake/functionspace.py

```diff
@@ -599,6 +599,45 @@ def __getitem__(self, i):
         return self
 
 
+class TensorFunctionSpace(FunctionSpaceBase):
+    """
+    A tensor-valued :class:`FunctionSpace`.
+    """
+    def __init__(self, mesh, family, degree, shape=None, symmetry=None, name=None,
+                 vfamily=None, vdegree=None):
+        if mesh.ufl_cell().cellname() == "quadrilateral":
+            raise NotImplementedError("Cannot currently build a TensorFunctionSpace on a quadrilateral mesh.")
+
+        if self._initialized:
+            return
+        mesh.init()
+        # TensorFunctionSpace shape defaults to the (gdim, gdim)
+        shape = shape or (mesh.ufl_cell().geometric_dimension(), )*2
+
+        if isinstance(mesh, mesh_t.ExtrudedMesh):
+            raise NotImplementedError("TFS on extruded meshes not implemented")
+        else:
+            element = ufl.TensorElement(family, domain=mesh.ufl_domain(),
+                                        degree=degree, shape=shape,
+                                        symmetry=symmetry)
+        super(TensorFunctionSpace, self).__init__(mesh, element, name,
+                                                  dim=shape, rank=2)
```

Comment:
Should rank always be 2, or should it be `len(shape)`.  I note a (somewhat cryptic) comment in `FunctionSpaceBase` says "this is the function space rank, the value rank may be different".  Do all tensor spaces have rank 2, irrespective of their shape?  Or does a TFS(...., shape=(1, 2, 3)) have rank 3?